src/parser.ts

```diff
@@ -4445,13 +4457,99 @@ export function parseImportExpression(
+  const node: ESTree.ImportExpression = {
+    type: 'ImportExpression',
+    source
+  };
+
+  if (context & Context.OptionsNext) {
+    let options: ESTree.Expression | null = null;
+
+    if (parser.getToken() === Token.Comma) {
+      consume(parser, context, Token.Comma);
+
+      if (parser.getToken() !== Token.RightParen) {
+        const expContext = (context | Context.DisallowIn) ^ Context.DisallowIn;
+        options = parseExpression(parser, expContext, 1, inGroup, parser.tokenPos, parser.linePos, parser.colPos);
+      }
+    }
+
+    node.options = options;
+    consumeOpt(parser, context, Token.Comma);
+  }
 
   consume(parser, context, Token.RightParen);
 
-  return finishNode(parser, context, start, line, column, {
-    type: 'ImportExpression',
-    source
-  });
+  return finishNode(parser, context, start, line, column, node);
+}
+
+/**
+ * Parses import attributes
+ *
+ * @param parser Parser object
+ * @param context Context masks
+ * @returns
+ */
+
+export function parseImportAttributes(
+  parser: ParserState,
+  context: Context,
+  specifiers: ESTree.ImportDeclaration['specifiers'] = []
+): ESTree.ImportAttribute[] {
+  consume(parser, context, Token.WithKeyword);
+  consume(parser, context, Token.LeftBrace);
+
+  const attributes: ESTree.ImportAttribute[] = [];
+  const keysContent = new Set<ESTree.Literal['value'] | ESTree.Identifier['name']>();
+  const validJSONImportAttributeBindings =
+    specifiers.length >= 1 &&
+    (specifiers[0].type === 'ImportDefaultSpecifier' ||
+      specifiers[0].type === 'ImportNamespaceSpecifier' ||
+      (specifiers[0].type === 'ImportSpecifier' && specifiers[0].imported.name === 'default'));
+
+  while (parser.getToken() !== Token.RightBrace) {
+    const start = parser.tokenPos;
+    const line = parser.linePos;
+    const column = parser.colPos;
+
+    const key = parseIdentifierOrLiteral(parser, context);
+    consume(parser, context, Token.Colon);
+    const value = parseLiteral(parser, context);
+    const keyContent = key.type === 'Literal' ? key.value : key.name;
+    const isJSONImportAttribute = keyContent === 'type' && value.value === 'json';
+
+    if (isJSONImportAttribute && !validJSONImportAttributeBindings) {
+      report(parser, Errors.InvalidImportBinding);
+    }
+
+    if (keysContent.has(keyContent)) {
+      report(parser, Errors.DuplicateBinding, `${keyContent}`);
+    }
+
+    keysContent.add(keyContent);
+    attributes.push(
+      finishNode(parser, context, start, line, column, {
+        type: 'ImportAttribute',
+        key,
+        value
+      })
+    );
+
+    if (parser.getToken() !== Token.RightBrace) {
+      consume(parser, context, Token.Comma);
+    }
+  }
+
+  consume(parser, context, Token.RightBrace);
+  return attributes;
+}
+
+function parseIdentifierOrLiteral(parser: ParserState, context: Context): ESTree.Identifier | ESTree.Literal {
```

Comment:
Suggest to rename it to parseIdentifierOrStringLiteral.
Do we have any test cases to reject?
```
with { 1: "foo" }
with { type: 1 }
with { type: [1] }
with { type: null }
with { type: undefined }
with { type: 'json', foo: {} }
...
```

Reply:
Looks like your implementation didn't reject the above cases.

From the proposal: 
```
We could permit import attributes to have more complex values than simply strings, for example:

import value from "module" with { attr: { key1: "value1", key2: [1, 2, 3] } };
This would allow import attributes to scale to support a larger variety of metadata.

We propose to omit this generalization in the initial proposal, as a key/value list of strings already affords significant flexibility to start, but we're open to a follow-on proposal providing this kind of generalization.
```

Reply:
test262 cases are way inadequate :-)